File: projects/cpp/DataSource/File/Read.cpp
/*
 * DataSource::File::Read
 * (c) 2023 Sam Caldwell.  See LICENSE.txt
 */
#include "File.h"
/*
 * Read a byte from the given position
 */
byte_t DataSource::File::Read(uint64_t position) {
    byte_t data;
    file.seekg(position);
    file.read(reinterpret_cast<char *>(&data), sizeof(byte_t));
    return data;
}

/*
 * Read a byte from the current position
 */
byte_t DataSource::File::Read() {
    byte_t data;
    file.read(reinterpret_cast<char *>(&data), sizeof(byte_t));
    return data;
}

File: go/crsce/BitFile/Read.go
package bitfile

/*
 * CRSCE BitFile reader
 * (c) 2023 Sam Caldwell.  See LICENSE.txt
 *
 * bit-for-bit reader/writer
 */

// Read - Read file o.file 4K at a time and return the buffer one bit at a time each call.
func (o *BitFile) Read() (bit bool, err error) {

	// adjust the bit position to the next bit or get the next byte in the buffer,
	// resetting the bit position to 0.
	if o.bitPos <= 8 {
		o.bitPos = 0
		o.bufferPos++
	}

	// if the buffer is exhausted, read more content into the buffer and
	// reset the buffer position to 0.
	if (o.bufferPos == bufferSize) || (o.buffer == nil) || (len(o.buffer) == 0) {
		o.buffer = make([]byte, bufferSize)
		o.bufferPos = 0
		o.bufferSize, err = o.file.Read(o.buffer)
		if err != nil {
			return false, err //Bail!
		}
	}

	// mask out only the needed bit.  if the result is not 0,
	// then return the set bit.
	mask := byte(1) << o.bitPos
	bit = (o.buffer[o.bufferPos] & mask) != 0

	// Return the bit
	return bit, nil
}
